core/src/layers/dtrace.rs

```diff
@@ -0,0 +1,440 @@
+/// Then you can use `readelf -n target/debug/examples/dtrace` to see the probes:
+///
+/// ```text
+/// Displaying notes found in: .note.stapsdt
+///   Owner                Data size        Description
+///   stapsdt              0x00000039       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: create_dir_start
+///     Location: 0x00000000000f8f05, Base: 0x0000000000000000, Semaphore: 0x00000000003649f8
+///     Arguments: -8@%rax
+///   stapsdt              0x00000037       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: create_dir_end
+///     Location: 0x00000000000f9284, Base: 0x0000000000000000, Semaphore: 0x00000000003649fa
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003c       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: blocking_list_start
+///     Location: 0x00000000000f9487, Base: 0x0000000000000000, Semaphore: 0x0000000000364a28
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003a       NT_STAPSDT (SystemTap probe descriptors)
+///     Provider: opendal
+///     Name: blocking_list_end
+///     Location: 0x00000000000f9546, Base: 0x0000000000000000, Semaphore: 0x0000000000364a2a
+///     Arguments: -8@%rax
+///   stapsdt              0x0000003c       NT_STAPSDT (SystemTap probe descriptors)
+/// ```
+#[derive(Default, Debug, Clone)]
+pub struct DtraceLayer {}
+
+impl<A: Accessor> Layer<A> for DtraceLayer {
+    type LayeredAccessor = DTraceAccessor<A>;
+    fn layer(&self, inner: A) -> Self::LayeredAccessor {
+        DTraceAccessor { inner }
+    }
+}
+
+#[derive(Clone)]
+pub struct DTraceAccessor<A: Accessor> {
+    inner: A,
+}
+
+impl<A: Accessor> Debug for DTraceAccessor<A> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DTraceAccessor")
+            .field("inner", &self.inner)
+            .finish_non_exhaustive()
+    }
+}
+
+#[async_trait]
+impl<A: Accessor> LayeredAccessor for DTraceAccessor<A> {
+    type Inner = A;
+    type Reader = DtraceLayerWarpper<A::Reader>;
+    type BlockingReader = DtraceLayerWarpper<A::BlockingReader>;
+    type Writer = DtraceLayerWarpper<A::Writer>;
+    type BlockingWriter = DtraceLayerWarpper<A::BlockingWriter>;
+    type Lister = A::Lister;
+    type BlockingLister = A::BlockingLister;
+
+    fn inner(&self) -> &Self::Inner {
+        &self.inner
+    }
+
+    async fn create_dir(&self, path: &str, args: OpCreateDir) -> Result<RpCreateDir> {
+        let c_path = CString::new(path).unwrap();
+        probe_lazy!(opendal, create_dir_start, c_path.as_ptr());
+        let result = self.inner.create_dir(path, args).await;
+        probe_lazy!(opendal, create_dir_end, c_path.as_ptr());
+        result
+    }
+
+    async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let c_path = CString::new(path).unwrap();
+        probe_lazy!(opendal, fetch_reader_start, c_path.as_ptr());
```

Comment:
We have formal names for all operation, please use them instead of designing new patterns.

[opendal::raw::Operation](https://opendal.apache.org/docs/rust/opendal/raw/enum.Operation.html) For all operations, and [opendal::raw::oio::ReadOperation](https://opendal.apache.org/docs/rust/opendal/raw/oio/enum.ReadOperation.html) for `reader` operations. `writer` and `lister` also have their own operation set.

For example:

- read could have `read_start` and `read_end`.
- reader's poll_read could have `reader_read_start`, `reader_read_pending` and `reader_read_end`.

Reply:
SGTM

Reply:
Another issue 

Should we keep `poll_read` and `read` to identify different event for oio::Reader and oio::BlockingReader? (writer as the same)

Reply:
We have ReadOperation::Read and ReadOperation::BlockingRead which I think is enough.